### PR TITLE
schedule, server: fix replicated state semantics

### LINKED ISSUE #11055: `/regions/replicated` reports incorrect placement scheduling state
### What did you do?

Configure placement rules for a key range and query:

`GET /pd/api/v1/regions/replicated?startKey=...&endKey=...`

Exercise the following states:

- All placement rules are satisfied.
- A missing peer has a legal scheduling target.
- A legal target is temporarily unavailable.
- A missing, down, or offline peer has no legal target.
- A Region has pending peers or no leader.
- The requested key range has no Region metadata or contains an internal gap.

This is related to #5095, which previously addressed inaccurate results caused
by the waiting list, but the endpoint still cannot reliably represent placement
scheduling progress.

### What did you expect to see?

- `REPLICATED` only when RuleChecker has no remaining placement work.
- `INPROGRESS` when placement is incomplete but can still progress, is waiting
  for temporary state recovery, or the requested range metadata is incomplete.
- `PENDING` only when the current topology has no legal placement action.

The status query should be side-effect free and should not require a full Store
scan for every completed Region.

### What did you see instead?

The endpoint combines placement fit results with state from the checker pending
list. The pending list represents historical or temporary filtering state, not
whether the current topology has a legal scheduling target.

As a result:

- Schedulable or temporarily blocked work can be reported as `PENDING`.
- Placement work with no legal target can remain `INPROGRESS`.
- An empty or discontinuous Region scan can be reported as complete.
- Reading the status can interact with placement-fit cache state.

### What version of PD are you using (`pd-server -V`)?

Current `master` at commit `070828c2b2032c499ff224c3f3730b5637fd3226`.

### PR DESCRIPTION
**Pull request by rleungx**
### What problem does this PR solve?

The `/regions/replicated` endpoint treats the RuleChecker pending list as the
final state. That list records a failed target selection and can become stale
after the Store topology changes, so schedulable work may still be reported as
`PENDING`. It can also report `REPLICATED` when Region metadata does not cover
the requested range.

Issue Number: Close #11055

### What is changed and how does it work?

```commit-message
Evaluate current placement state for every scanned Region without creating
Operators or mutating RuleChecker caches or metrics.

Return PENDING only when the current topology has no legal action, INPROGRESS
for schedulable or retryable work, and REPLICATED once no work remains. Reuse
a shared orphan-peer planner so the query matches scheduling behavior.

Keep healthy Regions on the fit-only fast path. For incomplete Regions, reuse
the Store list within the range check and use existence-only candidate scans
that stop at the first legal target. Inspect the running and waiting operator
sets under read locks, and validate complete Region coverage.
```

### Check List

Tests

- Unit test
- Integration test

Code changes

- Has HTTP API behavior changed

Side effects

- Healthy Regions stay on the fit-only fast path. For incomplete Regions,
  reuse the Store list within the range check and stop candidate scans at the
  first legal target.
- Operator inspection only takes read locks and does not maintain a per-Region
  waiting-operator index.

### Release note

```release-note
Fix inaccurate placement scheduling states returned by the regions replicated API.
```

## Summary by CodeRabbit

- **New Features**
  - Replication status checks now more precisely distinguish **PENDING** (blocked by unavailable placement/peers) from **INPROGRESS** (work can continue or retry), including gaps or incomplete region metadata.
  - Placement-rule-aware evaluation better reflects orphan peer handling, alternative-leader availability, and isolation/peer state constraints.

- **Bug Fixes**
  - Improved correctness of **replicated** vs progress states for ranges affected by missing data, scanned gaps, offline/down peers, and leader-rejection scenarios.
  - Updated API documentation for `/regions/replicated` to match the refined status meanings.

- **Tests**
  - Expanded coverage for the new status transitions and edge cases.

### CODE REVIEW

File: pkg/schedule/checker/rule_checker.go [hunk 1]
```diff
@@ -67,6 +80,403 @@ func (*RuleChecker) Name() string {
...
+func (c *RuleChecker) GetRegionPlacementState(region *core.RegionInfo) RegionPlacementState {
+	return c.evaluateRegionPlacementState(region, &placementStateContext{})
+}
+
+func (c *RuleChecker) evaluateRegionPlacementState(region *core.RegionInfo, context *placementStateContext) RegionPlacementState {
+	if region.GetLeader() == nil || len(region.GetPendingPeers()) > 0 {
+		return RegionPlacementStateInProgress
+	}
+	fit := c.ruleManager.FitRegionWithoutCache(c.cluster, region)
+	if isRegionPlacementSatisfied(region, fit) {
```
Comment (lhy1024) [inline pkg/schedule/checker/rule_checker.go:128]: **Blocking: this fast path skips real soft-location work.**

`isRegionPlacementSatisfied` only proves that the hard rule constraints and the configured `IsolationLevel` are satisfied. In particular, `IsRegionLabelIsolationSatisfied` always returns true when `IsolationLevel` is empty. Returning here therefore skips `hasBetterLocation`, even though the real RuleChecker may immediately create a `move-to-better-location` operator. This violates the stated meaning of `REPLICATED`: no placement work remains.

I verified this on `8e9a02f` by extending the existing `TestBetterReplacement` case:

```go
state := suite.rc.GetRegionPlacementState(suite.cluster.GetRegion(1))
op := suite.rc.Check(suite.cluster.GetRegion(1))
re.NotNil(op)
re.Equal("move-to-better-location", op.Desc())
re.Equal(uint64(4), op.Step(0).(operator.AddLearner).ToStore)
re.Equal(RegionPlacementStateInProgress, state)
```

All operator assertions pass, but the final assertion fails with `expected: 1 (INPROGRESS), actual: 0 (REPLICATED)`.

Please retain the fast path only when there are no location labels, or when the current placement is provably already optimal, such as being isolated at the highest location-label level. Otherwise the state path must evaluate better-location candidates.

File: pkg/schedule/checker/rule_checker.go [hunk 2]
```diff
@@ -67,6 +80,403 @@ func (*RuleChecker) Name() string {
...
+		}
+		oldRuleFit := fit.GetRuleFit(peer.GetId())
+		if oldRuleFit == nil || oldRuleFit == missingRuleFit || !oldRuleFit.IsSatisfied() {
+			continue
+		}
+
+		fastFailover := isWitnessEnabled(c.cluster) && store.IsTiKV() && oldRuleFit.Rule.IsWitness
+		strategy, candidateSet := context.getStrategy(c, region, oldRuleFit.Rule, fastFailover)
+		// addRulePeerWithOptions stops after trying the first matching peer.
+		return strategy.hasStoreToFix(candidateSet, getRuleFitStores(c.cluster, oldRuleFit), peer.GetStoreId())
```
Comment (lhy1024) [inline pkg/schedule/checker/rule_checker.go]: **Blocking: this deliberately mirrors a checker search-order dead end and reports `PENDING` even though the topology has a complete legal convergence path.**

In the test topology from `TestRegionPlacementFollowsSwapPeerOrder`, replacing the peer on Store 2 with Store 3 is legal. Once that move is applied, Store 2 is free and the checker can add the missing learner there. The generic checker currently tries the matching peer on Store 1 first, cannot replace it, and stops without trying Store 2. Returning after that first peer turns an implementation search-order limitation into the claim that the topology has no legal action.

I verified the full two-step path on `8e9a02f`:

```go
state := suite.rc.GetRegionPlacementState(region)
fit := suite.ruleManager.FitRegionWithoutCache(suite.cluster, region)
peer := region.GetStorePeer(2)
legalOp, err := suite.rc.replaceUnexpectedRulePeer(
	region, fit.GetRuleFit(peer.GetId()), fit, peer, "swap-fit",
)
re.NoError(err)
re.NotNil(legalOp)
re.Equal(uint64(3), legalOp.Step(0).(operator.AddLearner).ToStore)

re.Nil(suite.rc.Check(region)) // The generic path stops at Store 1.
movedRegion := region.Clone(
	core.WithRemoveStorePeer(2),
	core.WithAddPeer(&metapb.Peer{Id: 100, StoreId: 3, Role: metapb.PeerRole_Voter}),
)
nextOp := suite.rc.Check(movedRegion)
re.NotNil(nextOp)
re.Equal("add-rule-peer", nextOp.Desc())
re.Equal(uint64(2), nextOp.Step(0).(operator.AddLearner).ToStore)
re.Equal(RegionPlacementStateInProgress, state)
```

Every convergence-path assertion passes; only the final state assertion fails with `expected: 1 (INPROGRESS), actual: 2 (PENDING)`.

Please make the production checker continue through all matching peers, then share that action-selection result with the state query. The current test expectation of `PENDING` codifies the dead end and contradicts the PR contract that `PENDING` means no legal placement action exists.

File: pkg/schedule/checker/replica_strategy.go [hunk 3]
```diff
@@ -110,6 +184,57 @@ func (s *ReplicaStrategy) SelectStoreToFix(coLocationStores []*core.StoreInfo, o
...
+}
+
+func (s *ReplicaStrategy) hasBetterLocation(
+	candidateSet *storeCandidateSet,
+	cluster sche.SharedCluster,
+	region *core.RegionInfo,
+	fit *placement.RegionFit,
+	ruleFit *placement.RuleFit,
+) bool {
+	oldStoreID, _ := s.selectStoreToRemoveWithTempState(ruleFit.Stores)
```
Comment (lhy1024) [inline pkg/schedule/checker/replica_strategy.go]: **Blocking against the side-effect-free contract: this state query still mutates source-filter metrics.**

`hasBetterLocation` calls `selectStoreToRemoveWithTempState`. Its strict source pass uses `FilterSource(..., nil, nil, ...)`; when the counter argument is nil, `SelectSourceStores` increments the global `filterSourceCounter` for rejected Stores. Polling the GET endpoint therefore changes RuleChecker filter metrics whenever a source is busy, snapshot-limited, or remove-limited.

I verified this on `8e9a02f` with three peers in zone `z1`, a legal target in zone `z2`, hard zone isolation, and source Store 1 marked busy:

```go
before := sourceFilterMetricValue(re, suite.rc.Name(), "1")
re.Equal(
	RegionPlacementStateInProgress,
	suite.rc.GetRegionPlacementState(suite.cluster.GetRegion(1)),
)
after := sourceFilterMetricValue(re, suite.rc.Name(), "1")
re.Equal(before, after)
```

The state assertion passes, but the metric assertion fails with `expected: 0, actual: 1` for `pd_schedule_source_filter{scope="rule-checker",source="1"}` after a single query.

Please use a metric-free source predicate/selection helper for this read-only path, analogous to the direct target-filter evaluation added for candidate existence checks.

File: pkg/schedule/checker/rule_checker.go (hunk 1 above)
Comment (rleungx) [reply pkg/schedule/checker/rule_checker.go:128]: Fixed in 3289d20cc. The satisfied fast path now returns REPLICATED only when the same-role co-location set is already isolated at the highest configured location level. Otherwise the state query evaluates better-location candidates. Regression coverage includes both single-rule and multi-rule placement.

File: pkg/schedule/checker/rule_checker.go (hunk 2 above)
Comment (rleungx) [reply pkg/schedule/checker/rule_checker.go]: Fixed in 3289d20cc. Both the production RuleChecker and placement-state evaluator now continue through all eligible swap candidates. The regression test verifies the Store 2 → Store 3 move and the subsequent learner addition on Store 2. The same candidate-local dead end was also fixed for orphan, unhealthy-peer, and role-repair paths.

File: pkg/schedule/checker/replica_strategy.go (hunk 3 above)
Comment (rleungx) [reply pkg/schedule/checker/replica_strategy.go]: Fixed in 3289d20cc. The read-only state path now uses the shared source-selection algorithm without recording filter metrics, while the production scheduling path retains metric collection. A regression test verifies that state queries leave the metric unchanged and normal RuleChecker execution still increments it.
